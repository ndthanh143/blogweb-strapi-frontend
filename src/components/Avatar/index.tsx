import { useEffect, useState } from 'react';
import Image from 'next/image';
import { HiUserCircle } from 'react-icons/hi';

type HandleChange = (data: FileList | null) => void;

export type AvatarProps = {
  src?: string;
  width: number;
  height: number;
  alt: string;
  size: string;
  onClick?: () => void;
} & (({ isPicker: true } & { onChange: HandleChange }) | ({ isPicker?: false } & { onChange?: HandleChange }));

export function Avatar({ src, width, height, alt, size, isPicker, onClick, onChange }: AvatarProps) {
  const [file, setFile] = useState<File | null>(null);
  const [previewUrl, setPreviewUrl] = useState<string | null>(null);

  useEffect(() => {
    if (file) {
      const reader = new FileReader();
      reader.onloadend = () => {
        setPreviewUrl(reader.result as string);
      };
      reader.readAsDataURL(file);
    } else {
      setPreviewUrl(null);
    }
  }, [file]);

  return (
    <div
      className="rounded-full overflow-hidden cursor-pointer relative hover:opacity-60"
      style={{ width: `${width}px`, height: `${height}px` }}
      onClick={onClick}
    >
      {previewUrl ? (
        <Image
          src={previewUrl.replace(/\/v\d+\//g, '/q_40/')}
          fill
          quality={100}
          alt={alt}
          style={{ objectFit: 'cover', width: '100%', height: '100%' }}
          sizes={size}
        />
      ) : src ? (
        <Image
          src={src.replace(/\/v\d+\//g, '/q_40/')}
          fill
          quality={100}
          alt={alt}
          style={{ objectFit: 'cover', width: '100%', height: '100%' }}
          sizes={size}
          priority
        />
      ) : (
        <HiUserCircle className="w-full h-full text-gray-400 dark:text-white" />
      )}
      {isPicker && (
        <label
          htmlFor="imagePicker"
          className="absolute font-bold w-full h-full top-0 left-0 rounded-full flex items-center text-center hover:bg-gray-200 hover:bg-opacity-30 text-transparent hover:text-black cursor-pointer z-20"
        >
          Choose Image
          <input
            id="imagePicker"
            type="file"
            hidden
            onChange={(e) => {
              setFile(e.target.files && e.target.files?.[0]);
              onChange(e.target.files);
            }}
          />
        </label>
      )}
    </div>
  );
}
