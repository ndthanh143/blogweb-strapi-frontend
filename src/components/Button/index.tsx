import React, { ButtonHTMLAttributes, DetailedHTMLProps, PropsWithChildren } from 'react';
import cx from 'classnames';
import { AiOutlineLoading3Quarters } from 'react-icons/ai';

export interface ButtonProps extends DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  variant: 'solid' | 'outlined' | 'text';
  loading?: boolean;
  loadingPosition?: 'start' | 'end';
}

export function Button({
  variant,
  loading,
  loadingPosition = 'start',
  children,
  className,
  ...props
}: PropsWithChildren<ButtonProps>) {
  return (
    <button
      {...props}
      className={cx(
        variant === 'solid' &&
          'bg-color-primary text-white hover:bg-color-primary-hover active:color-primary-active flex justify-center items-center border-transparent',
        variant === 'outlined' &&
          'text-color-thin  dark:border-dark-mode dark:text-color-thin-dark hover:bg-gray-200 active:bg-gray-300 dark:hover:bg-blue-500 dark:hover:text-white dark:hover:border-blue-500 dark:active:bg-blue-700 dark:active:border-blue-700',
        variant === 'text' && 'hover:bg-gray-200 dark:hover:bg-slate-500 border-transparent',
        'py-2 px-4 rounded-md font-semibold border',
        loading && 'opacity-60 hover:bg-none active:bg-none',
        props.disabled && 'opacity-60',
        className,
      )}
    >
      {loading && loadingPosition == 'start' && <AiOutlineLoading3Quarters className="animate-spin-fast mx-2" />}
      {children}
      {loading && loadingPosition == 'end' && <AiOutlineLoading3Quarters className="animate-spin-fast mx-2" />}
    </button>
  );
}
