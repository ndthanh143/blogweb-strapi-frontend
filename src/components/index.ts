export * from './Avatar';
export * from './Button';
export * from './Card';
export * from './Comment';
export * from './Footer';
export * from './Header';
export * from './Input';
export * from './Label';
export * from './MiniNavigation';
export * from './Pagination';
export * from './Popper';
export * from './Select';
export * from './Slider';
export * from './SocialMedia';
export * from './SwitchMode';
export * from './Layout';
export * from './Seo';
export * from './Context';
