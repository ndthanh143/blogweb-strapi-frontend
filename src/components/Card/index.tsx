import Image from 'next/image';
import { HTMLAttributes, memo } from 'react';
import { Avatar, Label } from '@/components';
import cx from 'classnames';
import Link from 'next/link';
import moment from 'moment';
import { getStrapiMedia } from '@/utils/media';
import { Article } from '@/services/article/article.dto';
import { useRouter } from 'next/router';
import 'moment/locale/vi';

type FilterArticle = Omit<Article, 'content' | 'createdAt' | 'updatedAt' | 'description'>;

export type CardProps = (({ isLoading: true } & Partial<FilterArticle>) | ({ isLoading?: false } & FilterArticle)) & {
  isRendered?: boolean;
} & HTMLAttributes<HTMLDivElement>;

export const Card = memo(
  ({ title, thumbnail, slug, category, author, publishedAt, className, isLoading }: CardProps) => {
    const { locale } = useRouter();

    moment.locale(locale || 'en');

    return (
      <div
        className={cx(
          'w-full h-full p-4 border-2 border-light-mode dark:border-dark-mode rounded-lg hover:border-gray-300 dark:hover:border-gray-300',
          className,
        )}
      >
        <div className="aspect-[4/3] relative overflow-hidden rounded-lg">
          {isLoading ? (
            <div className="flex items-center justify-center h-full mb-4 bg-gray-300 rounded dark:bg-gray-700 animate-pulse">
              <svg
                className="w-12 h-12 text-gray-200 dark:text-gray-600"
                xmlns="http://www.w3.org/2000/svg"
                aria-hidden="true"
                fill="currentColor"
                viewBox="0 0 640 512"
              >
                <path d="M480 80C480 35.82 515.8 0 560 0C604.2 0 640 35.82 640 80C640 124.2 604.2 160 560 160C515.8 160 480 124.2 480 80zM0 456.1C0 445.6 2.964 435.3 8.551 426.4L225.3 81.01C231.9 70.42 243.5 64 256 64C268.5 64 280.1 70.42 286.8 81.01L412.7 281.7L460.9 202.7C464.1 196.1 472.2 192 480 192C487.8 192 495 196.1 499.1 202.7L631.1 419.1C636.9 428.6 640 439.7 640 450.9C640 484.6 612.6 512 578.9 512H55.91C25.03 512 .0006 486.1 .0006 456.1L0 456.1z" />
              </svg>
            </div>
          ) : (
            <Link
              href={`/blog/${slug}`}
              className="w-full h-full block relative"
              data-cy="post"
              aria-label={`Go to ${title}`}
            >
              {thumbnail.data && (
                <Image
                  src={getStrapiMedia(
                    thumbnail.data.attributes.formats.small ||
                      thumbnail.data.attributes.formats.thumbnail ||
                      thumbnail.data.attributes.formats.medium ||
                      thumbnail.data.attributes.formats.large,
                  )}
                  alt={thumbnail.data.attributes.alternativeText || ''}
                  fill
                  style={{ objectFit: 'cover' }}
                  className="hover:scale-110 transition-all duration-200"
                  sizes={thumbnail.data.attributes.formats.thumbnail + ''}
                  priority
                />
              )}
            </Link>
          )}
        </div>
        <div className="grid grid-rows-4 mt-4">
          <div className="row-span-1">
            {isLoading ? (
              <div className="h-8 bg-gray-200 rounded-full dark:bg-gray-700 w-32 mb-4 animate-pulse"></div>
            ) : (
              <Link href={`/category/${category.data.attributes.slug}`}>
                <Label color="secondary">{category.data.attributes.name}</Label>
              </Link>
            )}
          </div>
          {isLoading ? (
            <div className="row-span-2 animate-pulse">
              <div className="h-5 bg-gray-200 rounded-full dark:bg-gray-700 mb-2.5 "></div>
              <div className="h-5 bg-gray-200 rounded-full dark:bg-gray-700 mb-2.5 "></div>
              <div className="h-5 bg-gray-200 rounded-full dark:bg-gray-700"></div>
            </div>
          ) : (
            <p className="font-bold text-lg dark:text-color-bold-dark row-span-2">{title}</p>
          )}

          {isLoading ? (
            <div className="flex items-center">
              <svg
                className="text-gray-200 w-14 h-14 dark:text-gray-700"
                aria-hidden="true"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  fillRule="evenodd"
                  d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zm-2 4a5 5 0 00-4.546 2.916A5.986 5.986 0 0010 16a5.986 5.986 0 004.546-2.084A5 5 0 0010 11z"
                  clipRule="evenodd"
                ></path>
              </svg>
              <div>
                <div className="w-48 h-4 bg-gray-200 rounded-full dark:bg-gray-700"></div>
              </div>
            </div>
          ) : (
            author && (
              <div className="flex items-center font-sm text-color-blur row-span-1">
                <Link href={`/writer/${author.data.id}`} className="flex items-center">
                  <Avatar
                    src={
                      author.data.attributes.avatar?.data &&
                      getStrapiMedia(author.data.attributes.avatar.data.attributes.formats.thumbnail)
                    }
                    width={40}
                    height={40}
                    alt={author.data.attributes.avatar?.data?.attributes.alternativeText || ''}
                    size={author.data.attributes.avatar?.data?.attributes.formats.thumbnail + ''}
                  />
                  <p className="ml-2 mr-4 font-medium hover:text-color-primary">{author.data.attributes.name}</p>
                </Link>
                <p>{moment(publishedAt).format(locale === 'vi' ? 'DD MMMM, YYYY' : 'MMMM DD, YYYY')}</p>
              </div>
            )
          )}
        </div>
      </div>
    );
  },
  (prevProps, nextProps) => {
    return prevProps.isRendered === nextProps.isRendered;
  },
);
