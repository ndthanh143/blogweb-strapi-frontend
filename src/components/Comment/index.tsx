import { CommentAttribute, UpdateCommentPayload } from '@/services/comment/comment.dto';
import { Avatar, Button, Popper } from '@/components';
import { getStrapiMedia } from '@/utils/media';
import moment from 'moment';
import Link from 'next/link';
import { FaAngleDown, FaAngleUp, FaPencilAlt, FaTrash } from 'react-icons/fa';
import { HiOutlineDotsCircleHorizontal } from 'react-icons/hi';
import useBoolean from '@/hooks/useBoolean';
import { useAuth } from '@/hooks/useAuth';
import { HTMLAttributes, useEffect, useState } from 'react';
import dynamic from 'next/dynamic';
import { useAppSelector } from '@/redux/store';
import cx from 'classnames';
import Markdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
const Editor = dynamic(() => import('@/components/Editor/Editor'), { ssr: false });

export interface ICommentProps extends HTMLAttributes<HTMLDivElement> {
  data: CommentAttribute;
  onDelete: () => void;
  onUpdate: (comment: string) => void;
  onAnswer?: (reply: string) => void;
}

export function Comment({ data, onDelete = () => {}, onUpdate, onAnswer = () => {}, className }: ICommentProps) {
  const { user } = useAuth();

  const { value: isAction, setFalse: closeAction, toggle: toggleAction } = useBoolean(false);

  const { value: isUpdating, setFalse: closeUpdate, setTrue: openUpdate } = useBoolean(false);

  const { value: isAnswering, setFalse: closeAnswer, setTrue: openAnswer } = useBoolean(false);

  const [comment, setComment] = useState(data.content);

  const [reply, setReply] = useState('');

  const { isUpdateSuccess, isAnswerSuccess, updateLoading, answerLoading } = useAppSelector(
    (state) => state.handleComment,
  );

  useEffect(() => {
    if (isUpdateSuccess) {
      closeUpdate();
    }
    if (isAnswerSuccess) {
      closeAnswer();
      setReply('');
    }
  }, [isUpdateSuccess, isAnswerSuccess, closeUpdate, closeAnswer]);

  return (
    <div className={cx('border-b dark:border-dark-mode mb-2 last:mb-0 last:border-0', className)}>
      <div className="flex items-center mx-2">
        <Link href={`/writer/${data.user.data.id}`}>
          <Avatar
            src={
              data.user.data.attributes.avatar?.data &&
              getStrapiMedia(data.user.data.attributes.avatar.data.attributes.formats.thumbnail)
            }
            width={40}
            height={40}
            alt={data.user.data.attributes.avatar?.data?.attributes.alternativeText || ''}
            size={data.user.data.attributes.avatar?.data?.attributes.formats.thumbnail + '' || ''}
          />
        </Link>
        <div className="mx-2">
          <Link
            href={`/writer/${data.user.data.id}`}
            className="text-color-bold dark:text-color-bold-dark font-semibold hover:underline hover:decoration-1"
          >
            {data.user.data.attributes.name}
          </Link>
          <p className="text-color-thin text-sm">{moment(data.publishedAt).fromNow()}</p>
        </div>
      </div>

      {isUpdating && (
        <div className="py-2">
          <Editor value={comment} onChange={setComment} />
          <div className="flex flex-row-reverse my-4">
            <Button
              variant="solid"
              onClick={() => onUpdate(comment)}
              loading={updateLoading}
              disabled={updateLoading}
              aria-label="Save - Click to update your comment"
            >
              Save
            </Button>
            <Button variant="text" className="mx-2" onClick={closeUpdate} aria-label="Cancel">
              Cancel
            </Button>
          </div>
        </div>
      )}
      {!isUpdating && (
        <>
          <div className="text-lg font-light my-2">
            <section className="rich-text py-6">
              <Markdown remarkPlugins={[remarkGfm]}>{data.content}</Markdown>
            </section>
          </div>
          <div className="flex">
            <div className="flex items-center text-xl text-color-thin font-semibold border-r-2 dark:border-dark-mode px-2">
              <span className="opacity-40 hover:!opacity-100 cursor-pointer">
                <FaAngleUp />
              </span>
              <span className="opacity-70 text-lg">0</span>
              <span className="opacity-40 hover:!opacity-100 cursor-pointer">
                <FaAngleDown />
              </span>
            </div>
            <div className="flex items-center ml-2 ">
              <p className="text-blue-500 hover:underline hover:decoration-1 cursor-pointer" onClick={openAnswer}>
                Answer
              </p>
              {user && user.id === data.user.data.id && (
                <span className="text-color-thin mx-2 text-2xl cursor-pointer relative" onClick={toggleAction}>
                  <HiOutlineDotsCircleHorizontal />
                  <Popper
                    isOpen={isAction}
                    onClose={closeAction}
                    onItemClick={closeAction}
                    className="left-0 origin-top-left"
                  >
                    <span
                      className="cursor-pointer text-base hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white pl-4 py-2 flex items-center pr-8"
                      onClick={openUpdate}
                    >
                      <FaPencilAlt /> <span className="px-4">Update</span>
                    </span>
                    <span
                      className="cursor-pointer text-base hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white pl-4 py-2 flex items-center pr-8"
                      onClick={onDelete}
                    >
                      <FaTrash /> <span className="px-4">Remove this comment</span>
                    </span>
                  </Popper>
                </span>
              )}
            </div>
          </div>
          {isAnswering && (
            <div className="py-2">
              <Editor value={reply} onChange={setReply} />
              <div className="flex flex-row-reverse my-4">
                <Button
                  variant="solid"
                  onClick={() => onAnswer(reply)}
                  loading={answerLoading}
                  disabled={answerLoading}
                  aria-label="Reply - Click to submit your reply"
                >
                  Reply
                </Button>
                <Button variant="text" className="mx-2" onClick={closeAnswer} aria-label="Cancel">
                  Cancel
                </Button>
              </div>
            </div>
          )}
        </>
      )}
    </div>
  );
}
