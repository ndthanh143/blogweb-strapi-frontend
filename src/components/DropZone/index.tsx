import cx from 'classnames';
import { useEffect, useState } from 'react';
import Image from 'next/image';
import { HandleChange } from '../Avatar';

export interface DropZoneProps {
  onChange: HandleChange;
}

export function DropZone({ onChange, ...props }: DropZoneProps) {
  const [file, setFile] = useState<File | null>(null);
  const [previewUrl, setPreviewUrl] = useState<string | null>(null);

  useEffect(() => {
    if (file) {
      const reader = new FileReader();
      reader.onloadend = () => {
        setPreviewUrl(reader.result as string);
      };
      reader.readAsDataURL(file);
    } else {
      setPreviewUrl(null);
    }
  }, [file]);

  return (
    <div className="rounded-lg overflow-hidden justify-center w-full h-64 relative cursor-pointer">
      {previewUrl && (
        <Image
          src={previewUrl}
          fill
          quality={100}
          alt={'preview'}
          style={{ objectFit: 'cover', width: '100%', height: '100%' }}
          sizes={''}
        />
      )}
      <label
        htmlFor="dropzone-file"
        className={cx(
          ' absolute flex flex-col items-center justify-center w-full h-full border-2 border-gray-300 border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-bray-800 dark:bg-gray-700 hover:bg-gray-100 dark:border-gray-600 dark:hover:border-gray-500 dark:hover:bg-gray-600',
          previewUrl && 'opacity-0 hover:opacity-60',
        )}
      >
        <div className="flex flex-col items-center justify-center pt-5 pb-6">
          <svg
            aria-hidden="true"
            className="w-10 h-10 mb-3 text-gray-400"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M7 16a4 4 0 01-.88-7.903A5 5 0 1115.9 6L16 6a5 5 0 011 9.9M15 13l-3-3m0 0l-3 3m3-3v12"
            ></path>
          </svg>
          <p className="mb-2 text-sm text-gray-500 dark:text-gray-400">
            <span className="font-semibold">Click to upload</span> or drag and drop
          </p>
          <p className="text-xs text-gray-500 dark:text-gray-400">SVG, PNG, JPG or GIF (MAX. 800x400px)</p>
        </div>
        <input
          id="dropzone-file"
          type="file"
          className="hidden"
          onChange={(e) => {
            setFile(e.target.files && e.target.files?.[0]);
            onChange(e.target.files);
          }}
          {...props}
        />
      </label>
    </div>
  );
}
