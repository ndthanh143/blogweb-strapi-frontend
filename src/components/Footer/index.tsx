import { FiMail } from 'react-icons/fi';
import { useEffect } from 'react';
import Link from 'next/link';
import { useAppDispatch, useAppSelector } from '@/redux/store';
import { getCategories } from '@/redux/features/categories/categoriesSlice';
import { useTranslation } from 'next-i18next';
import { Button, Input } from '@/components';
import Logo from '@/assets/logo';

export function Footer() {
  const { t } = useTranslation('footer');

  const dispatch = useAppDispatch();

  const { data } = useAppSelector((state) => state.categories);

  const translate = {
    about: t('about'),
    quickLink: t('quickLink'),
    category: t('category'),
    home: t('home'),
    blog: t('blog'),
    terms: t('terms'),
    privacy: t('privacy'),
    cookie: t('cookie'),
  };

  useEffect(() => {
    if (data.length === 0) {
      dispatch(getCategories());
    }
  }, [data.length, dispatch]);

  return (
    <div className="bg-footer-color dark:bg-footer-color-dark">
      <div className="container mx-auto">
        <div className="grid lg:grid-cols-6 grid-cols-1 text-center lg:text-left place-items-stretch gap-x-20 border-b-2 dark:border-dark-mode py-6">
          <div className="text-color-medium dark:text-color-medium-dark text-sm lg:col-span-2 py-4">
            <h1 className="text-color-bold dark:text-color-bold-dark text-base font-bold mb-3">{translate.about}</h1>
            <p className="mb-3 text-color-thin dark:text-color-thin-dark">
              We are the best team and company in the Asia, join with us is the best choice and decision of your life,
              make it clearly and don&apos;t worry a bout monney. We are everythings you need
            </p>
            <p className="dark:text-color-thin-dark">
              <span className="text-base font-bold text-color-bold dark:text-color-bold-dark">Email: </span>
              alex.nguyen.goldenowl@gmail.com
            </p>
            <p className="dark:text-color-thin-dark">
              <span className="text-base font-bold text-color-bold dark:text-color-bold-dark">Phone: </span>
              +84 35 456 0042
            </p>
          </div>
          <div className="lg:col-span-1 text-color-medium dark:text-color-medium-dark text-sm py-4">
            <h1 className="text-color-bold dark:text-color-bold-dark text-base font-bold mb-3">
              {translate.quickLink}
            </h1>
            <Link href="/" className="mb-2 hover:text-red-500 block">
              {translate.home}
            </Link>
            <Link href={`/category/${data?.[0]?.attributes.slug}`} className="mb-2 hover:text-red-500 block">
              {translate.blog}
            </Link>
          </div>
          <div className="lg:col-span-1 text-color-medium dark:text-color-medium-dark text-sm py-4">
            <h1 className="text-color-bold dark:text-color-bold-dark text-base font-bold mb-3">{translate.category}</h1>
            {data.length > 0 &&
              data.map((item) => (
                <Link
                  href={`/category/${item.attributes.slug}`}
                  className="mb-2 hover:text-red-500 block w-full"
                  key={item.id}
                >
                  {item.attributes.name}
                </Link>
              ))}
          </div>
          <div className="lg:col-span-2 bg-white text-center px-6 rounded-lg dark:bg-input-dark py-4">
            <h1 className="font-bold mb-1 dark:text-color-bold-dark">Weekly Newsletter</h1>
            <p className="mb-8 dark:text-color-blur">Get blog articles and offers via email</p>
            <Input placeholder="Your Email" endDecorator={<FiMail />} />
            <Button variant="solid" type="submit" className="mt-3 w-full" aria-label="Subcribe Email">
              Subcribe
            </Button>
          </div>
        </div>
        <div className="grid lg:grid-cols-3 text-center lg:text-left items-center py-6">
          <div className="flex justify-center lg:col-span-2 lg:justify-start py-4">
            <Logo />
          </div>
          <div className="lg:col-span-1 lg:flex justify-between text-sm text-color-medium dark:text-color-medium-dark">
            <Link href="/" className="px-4 py-4 hover:text-red-500">
              {translate.terms}
            </Link>
            <Link href="/" className="px-4 py-4 hover:text-red-500">
              {translate.privacy}
            </Link>
            <Link href="/" className="px-4 py-4 hover:text-red-500">
              {translate.cookie}
            </Link>
          </div>
        </div>
      </div>
    </div>
  );
}
