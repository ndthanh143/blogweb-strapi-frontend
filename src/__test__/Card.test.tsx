import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import axios from 'axios';
import { Card } from '@/components';
import { ArticlesResponse } from '@/services/article/article.dto';
import { Router, useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
  Router: jest.fn(),
}));

describe('Card component', () => {
  test('should render correctly', async () => {
    const { data } = await axios.get<ArticlesResponse>('https://blogsite-backend-6e79.onrender.com/api/articles', {
      params: { populate: '*' },
    });
    const article = data.data[0].attributes;

    const component = renderer.create(
      <Card
        title={article.title}
        thumbnail={article.thumbnail}
        slug={article.slug}
        category={article.category}
        author={article.author}
        publishedAt={article.publishedAt}
      />,
    );
    let tree = component.toJSON() as ReactTestRendererJSON;
    expect(tree).toMatchSnapshot();
  });
});
