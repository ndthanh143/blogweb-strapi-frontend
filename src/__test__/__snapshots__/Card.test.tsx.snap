// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly 1`] = `
<div
  className="w-full h-full p-4 border-2 border-light-mode dark:border-dark-mode rounded-lg hover:border-gray-300 dark:hover:border-gray-300"
>
  <div
    className="aspect-[4/3] relative overflow-hidden rounded-lg"
  >
    <a
      className="w-full h-full block relative"
      data-cy="post"
      href="/blog/chuong-10-selection-algorithms-2-problems-and-solutions-13-21"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="what-s-inside-a-black-hole"
        className="hover:scale-110 transition-all duration-200"
        data-nimg="fill"
        decoding="async"
        fetchpriority="high"
        onError={[Function]}
        onLoad={[Function]}
        sizes="[object Object]"
        src="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=3840&q=75"
        srcSet="/_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fres.cloudinary.com%2Fdk9jp2gb5%2Fimage%2Fupload%2Fv1687419247%2Fsmall_what_s_inside_a_black_hole_360c63e36f.jpg&w=3840&q=75 3840w"
        style={
          {
            "bottom": 0,
            "color": "transparent",
            "height": "100%",
            "left": 0,
            "objectFit": "cover",
            "objectPosition": undefined,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": "100%",
          }
        }
      />
    </a>
  </div>
  <div
    className="grid grid-rows-4 mt-4"
  >
    <div
      className="row-span-1"
    >
      <a
        href="/category/technology"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <span
          className="px-4 py-1 rounded-lg font-medium bg-label-light text-color-primary dark:bg-label-dark hover:bg-color-primary dark:hover:bg-color-primary hover:text-white"
        >
          Technology
        </span>
      </a>
    </div>
    <p
      className="font-bold text-lg dark:text-color-bold-dark row-span-2"
    >
      Chương 10: SELECTION ALGORITHMS - 2.Problems & Solutions(13-21)
    </p>
    <div
      className="flex items-center font-sm text-color-blur row-span-1"
    >
      <a
        className="flex items-center"
        href="/writer/8"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="rounded-full overflow-hidden cursor-pointer relative hover:opacity-60"
          style={
            {
              "height": "40px",
              "width": "40px",
            }
          }
        >
          <svg
            aria-hidden="true"
            className="w-full h-full text-gray-400 dark:text-white"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clipRule="evenodd"
              d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zm-2 4a5 5 0 00-4.546 2.916A5.986 5.986 0 0010 16a5.986 5.986 0 004.546-2.084A5 5 0 0010 11z"
              fillRule="evenodd"
            />
          </svg>
        </div>
        <p
          className="ml-2 mr-4 font-medium hover:text-color-primary"
        >
          Kevin Nguyen
        </p>
      </a>
      <p>
        June 23, 2023
      </p>
    </div>
  </div>
</div>
`;
